### PR TITLE
feat: OAuth provider for Zapier

### LINKED ISSUE #9941: [CAL-2102] OAuth instead of API keys
ideally we implement some sort of OAuth á la "Continue with [Cal.com](http://Cal.com)" or "Sign in with [Cal.com](http://Cal.com)" to replace the manual API key setup for integration partners:

[Source image](https://github.com/calcom/cal.com/assets/8019099/eb65f878-7c9b-4c7f-bb39-3d74884b520d)

this will also make integrating with other apps a lot easier and faster (and more robust) as we dont rely on user action to generate and paste API keys.

we should also have some sort of "permission splash screen" that outlines what access the third party app will have

[Source image](https://github.com/calcom/cal.com/assets/8019099/a8a47e36-304f-4a84-a15e-e4a2494b6699)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2102](https://linear.app/calcom/issue/CAL-2102/oauth-instead-of-api-keys)</sub>

### PR DESCRIPTION
**Pull request by CarinaWolli**
## What does this PR do?

Implements the first version of the cal.com OAuth provider for Zapier. 

Fixes #9941 

Loom Video: https://www.loom.com/share/3b68afbf73a0400c94fe389acee4ab65

## Type of change

- [x] New feature

## How should this be tested?

- Sign in as admin and go to http://localhost:3000/settings/admin/oAuth
- Add new oAuth client and copy client id and secret to your notes 
- Go to `locahost:3000/auth/oauth2/authorize?client_id=<clientId>&client_secret=<clientSecret>&state=anyString&redirect_uri=<any url here></span>&scope=READ_PROFILE`

  - If not logged in you need to log in with any user and you will be redirected back to the link
- Click ‘Allow’ and you will be redirected to the redirect_uri with `code` and the `state` as params
  - e.g.: `<redirect_uri>/?code=OBW-JlonLV-lxYVF0Er1XOFtaG8O1l9-IhWN7h8Z9hXicaBmWrBCeQ&state=anyString`
- POST request to `<url>/api/auth/oauth/token` endpoint with the following body: (be aware that access code expires in 10 mins)
```
{
    "grant_type": "authorization_code",    
	"client_id": 	“<clientId>”,  
	“client_secret: 	“<clientSecret>”, 	
	"redirect_uri": "<redirectUri>",
    "code": "OBW-JlonLV-lxYVF0Er1XOFtaG8O1l9-IhWN7h8Z9hXicaBmWrBCeQ" —> from above
}
```
As a response you will get `acces_token` and `refresh_token` 

- Test GET request to `<url>/api/auth/oauth/me` endpoint with the access token in Authorization like that:
    - Bearer '<accessToken>'
- Test POST request to `<url>/api/auth/oauth/refreshToken` to request a new access token, with the following body: 
    - In the Authorization header add the refresh token as a Bearer token

```
{
    "grant_type": "refresh_token",    
    "client_id": “<clientId>”,    
    "client_secret": “<clientSecret>”, 
}
```

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

File: packages/trpc/server/routers/viewer/oAuth/generateAuthCode.handler.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  const team = isTeam ? await prisma.team.findFirst({ where: { slug: teamSlug.substring(5) } }) : undefined;
+
+  await prisma.accessCode.create({
+    data: {
+      code: authorizationCode,
+      clientId,
+      userId: !isTeam ? ctx.user.id : undefined,
+      teamId: team ? team.id : undefined,
+      expiresAt: dayjs().add(10, "minutes").toDate(),
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/oAuth/generateAuthCode.handler.ts:47]: Client exchanges the access code with the token (`auth/oAuth/token`), access code expires after 10 minutes

**Comment by PeerRich** [comment]
great work! really excited about this.

File: apps/web/playwright/oauth-provider.e2e.ts [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    await page.goto(
+      `auth/oauth2/authorize?client_id=${client.clientId}&redirect_uri=${client.redirectUri}&response_type=code&scope=READ_PROFILE&state=1234`
+    );
+
+    // check if user is redirected to login page
+    await expect(page.getByRole("heading", { name: "Welcome back" })).toBeVisible();
+    await page.locator("#email").fill(user.email);
+    await page.locator("#password").fill(user.username || "");
+
+    await page.getByRole("button", { name: "Sign in" }).click();
```
Comment (CarinaWolli) [inline apps/web/playwright/oauth-provider.e2e.ts:206]: this did pass locally, I will take a look why this fails now

Comment (CarinaWolli) [reply apps/web/playwright/oauth-provider.e2e.ts]: fixed

File: apps/web/pages/auth/oauth2/authorize.tsx [hunk 3]
```diff
@@ -0,0 +1,170 @@
+import { useSession } from "next-auth/react";
+import { useRouter } from "next/router";
```
Comment (sean-brydon) [inline apps/web/pages/auth/oauth2/authorize.tsx:2]:
```suggestion
import { useRouter } from "next/navigation";
```

File: apps/web/pages/auth/oauth2/authorize.tsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+import { Plus, Info } from "@calcom/ui/components/icon";
+
+import PageWrapper from "@components/PageWrapper";
+
+export default function Authorize() {
+  const { t } = useLocale();
+  const { status } = useSession();
+
+  const router = useRouter();
+  const { state, client_id, scope } = router.query;
```
Comment (sean-brydon) [inline apps/web/pages/auth/oauth2/authorize.tsx:18]: These should come from useSearchParams()

File: apps/web/pages/auth/oauth2/authorize.tsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    return <></>;
+  }
+
+  if (status === "unauthenticated") {
+    const urlSearchParams = new URLSearchParams({
+      callbackUrl: `auth/oauth2/authorize?${queryString}`,
+    });
+    router.replace(`/auth/login?${urlSearchParams.toString()}`);
+    return <></>;
+  }
```
Comment (sean-brydon) [inline apps/web/pages/auth/oauth2/authorize.tsx:68]: I think this should live in a useEffect 

```suggestion
useEffect(() => {
  if (status === "unauthenticated") {
    const urlSearchParams = new URLSearchParams({
      callbackUrl: `auth/oauth2/authorize?${queryString}`,
    });
    router.replace(`/auth/login?${urlSearchParams.toString()}`);
  }
}, [status]);
```

File: apps/web/pages/settings/admin/oAuth/oAuthView.tsx [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+};
+
+export default function OAuthView() {
+  const oAuthForm = useForm<FormValues>();
+  const [clientSecret, setClientSecret] = useState("");
+  const [clientId, setClientId] = useState("");
+  const [logo, setLogo] = useState("");
+
+  const mutation = trpc.viewer.oAuth.addClient.useMutation({
+    onSuccess: async (data: any) => {
```
Comment (sean-brydon) [inline apps/web/pages/settings/admin/oAuth/oAuthView.tsx:22]:
```suggestion
    onSuccess: async (data) => {
```
data should be typed from trpc - no need to set as any

File: apps/web/pages/settings/admin/oAuth/oAuthView.tsx [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+      setClientSecret(data.clientSecret);
+      setClientId(data.clientId);
+      showToast(`Successfully added ${data.name} as new client`, "success");
+    },
+    onError: (error: any) => {
```
Comment (sean-brydon) [inline apps/web/pages/settings/admin/oAuth/oAuthView.tsx:27]:
```suggestion
    onError: (error) => {
```
error should be typed from trpc - no need to set as any

File: apps/web/pages/settings/admin/oAuth/oAuthView.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                    }}
+                    type="button"
+                    className="rounded-l-none text-base"
+                    StartIcon={Clipboard}>
+                    Copy
+                  </Button>
+                </Tooltip>
+              </div>
+              <div className="text-subtle text-sm">
+                After copying the secret you won&apos;t be able to view it anymore
```
Comment (sean-brydon) [inline apps/web/pages/settings/admin/oAuth/oAuthView.tsx:132]: Missing i18n key

File: packages/app-store/zapier/api/subscriptions/addSubscription.ts [hunk 9]
```diff
@@ -1,26 +1,41 @@
...
 import findValidApiKey from "@calcom/features/ee/api-keys/lib/findValidApiKey";
 import { addSubscription } from "@calcom/features/webhooks/lib/scheduleTrigger";
 import { defaultHandler, defaultResponder } from "@calcom/lib/server";
 
 async function handler(req: NextApiRequest, res: NextApiResponse) {
   const apiKey = req.query.apiKey as string;
 
-  if (!apiKey) {
-    return res.status(401).json({ message: "No API key provided" });
+  let validKey: any = null;
```
Comment (sean-brydon) [inline packages/app-store/zapier/api/subscriptions/addSubscription.ts:11]: Havent tested code still performs the same way but i think we can remove the any and null define here. But on a type level we can improve this 

```suggestion
  let validKey;
```
CleanShot 2023-09-26 at 12 21 01

File: packages/app-store/zapier/api/subscriptions/deleteSubscription.ts [hunk 10]
```diff
@@ -13,18 +14,34 @@ const querySchema = z.object({
 async function handler(req: NextApiRequest, res: NextApiResponse) {
   const { apiKey, id } = querySchema.parse(req.query);
 
-  if (!apiKey) {
-    return res.status(401).json({ message: "No API key provided" });
+  const scopes = ["READ_BOOKING, READ_PROFILE"];
+
+  let validKey: any = null;
```
Comment (sean-brydon) [inline packages/app-store/zapier/api/subscriptions/deleteSubscription.ts:19]:
```suggestion
  let validKey;
```

Same as above

File: packages/app-store/zapier/api/subscriptions/listBookings.ts [hunk 11]
```diff
@@ -1,29 +1,54 @@
...
 import findValidApiKey from "@calcom/features/ee/api-keys/lib/findValidApiKey";
 import { listBookings } from "@calcom/features/webhooks/lib/scheduleTrigger";
 import { defaultHandler, defaultResponder } from "@calcom/lib/server";
 
 async function handler(req: NextApiRequest, res: NextApiResponse) {
   const apiKey = req.query.apiKey as string;
 
-  if (!apiKey) {
-    return res.status(401).json({ message: "No API key provided" });
+  let validKey: any = null;
```
Comment (sean-brydon) [inline packages/app-store/zapier/api/subscriptions/listBookings.ts:11]:
```suggestion
  let validKey;
```

File: packages/app-store/zapier/api/subscriptions/addSubscription.ts (hunk 9 above)
Comment (sean-brydon) [reply packages/app-store/zapier/api/subscriptions/addSubscription.ts:11]: If we need this to be null and not undefined ignore this

File: apps/web/pages/settings/admin/oAuth/oAuthView.tsx (hunk 8 above)
Comment (CarinaWolli) [reply apps/web/pages/settings/admin/oAuth/oAuthView.tsx:132]: I am missing more translations in this view, will fix that. Thank you 🙏

**Review by sean-brydon — COMMENTED** [review]
Wow @CarinaWolli what an awesome PR. Great job! 

Few NITs - nothing blocking i dont think

File: packages/features/webhooks/lib/scheduleTrigger.ts [hunk 12]
```diff
@@ -16,18 +16,27 @@ export async function addSubscription({
...
+    const userId = appApiKey ? appApiKey.userId : account && !account.isTeam ? account.id : null;
+    const teamId = appApiKey ? appApiKey.teamId : account && account.isTeam ? account.id : null;
+
     const createSubscription = await prisma.webhook.create({
       data: {
         id: v4(),
-        userId: appApiKey.userId,
-        teamId: appApiKey.teamId,
+        userId: userId,
+        teamId: teamId,
```
Comment (zomars) [inline packages/features/webhooks/lib/scheduleTrigger.ts:39]:
```suggestion
        userId,
        teamId,
```

File: packages/trpc/server/routers/viewer/oAuth/generateAuthCode.handler.ts [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+      name: true,
+    },
+  });
+
+  if (!client) {
+    throw new TRPCError({ code: "UNAUTHORIZED", message: "Client ID not valid" });
+  }
+  const authorizationCode = generateAuthorizationCode();
+
+  const isTeam = teamSlug.startsWith("team/");
```
Comment (zomars) [inline packages/trpc/server/routers/viewer/oAuth/generateAuthCode.handler.ts:37]: Why do we need to do this? Just trying to understand when this is the case.

File: packages/app-store/zapier/api/subscriptions/addSubscription.ts [hunk 14]
```diff
@@ -1,26 +1,15 @@
...
-
-  const validKey = await findValidApiKey(apiKey, "zapier");
-
-  if (!validKey) {
-    return res.status(401).json({ message: "API key not valid" });
-  }
-
   const { subscriberUrl, triggerEvent } = req.body;
-
+  const { account, appApiKey } = await validateAccountOrApiKey(req, ["READ_BOOKING", "READ_PROFILE"]);
```
Comment (zomars) [inline packages/app-store/zapier/api/subscriptions/addSubscription.ts:9]: Added some DRYness in here. @CarinaWolli hope you don't mind 🙏🏽

File: packages/app-store/zapier/api/subscriptions/listBookings.ts [hunk 15]
```diff
@@ -1,29 +1,31 @@
...
-  const validKey = await findValidApiKey(apiKey, "zapier");
-
-  if (!validKey) {
-    return res.status(401).json({ message: "API key not valid" });
-  }
-
-  const bookings = await listBookings(validKey);
+  const { account: authorizedAccount, appApiKey: validKey } = await validateAccountOrApiKey(req, [
+    "READ_BOOKING",
+  ]);
```
Comment (zomars) [inline packages/app-store/zapier/api/subscriptions/listBookings.ts:10]: Same here

File: packages/app-store/zapier/lib/validateAccountOrApiKey.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+  if (!apiKey) {
+    const authorizedAccount = await isAuthorized(req, requiredScopes);
+    if (!authorizedAccount) throw new HttpError({ statusCode: 401, message: "Unauthorized" });
+    return { account: authorizedAccount, appApiKey: undefined };
+  }
+
+  const validKey = await findValidApiKey(apiKey, "zapier");
+  if (!validKey) throw new HttpError({ statusCode: 401, message: "API key not valid" });
+  return { account: null, appApiKey: validKey };
+}
```
Comment (zomars) [inline packages/app-store/zapier/lib/validateAccountOrApiKey.ts:19]: Extracted this reused logic and de-nested it a little bit.

**Review by zomars — DISMISSED** [review]
I think this is in merge able state. Demo looks fine, tested locally. And no code complaints so far. We would need to test of QA cc @shivamklr 

Awesome work @CarinaWolli 🙏🏽

File: packages/trpc/server/routers/viewer/oAuth/generateAuthCode.handler.ts (hunk 13 above)
Comment (CarinaWolli) [reply packages/trpc/server/routers/viewer/oAuth/generateAuthCode.handler.ts]: Clean up the code a bit, I am now already making sure before that `teamSlug` only has the team slug and not also a username (https://github.com/calcom/cal.com/pull/11465/files#diff-77fc6507ac0ac4f357ba8d738052d1d61309bfc25c643599cfde2f344eefa565R165)

File: packages/app-store/zapier/api/subscriptions/addSubscription.ts (hunk 14 above)
Comment (CarinaWolli) [reply packages/app-store/zapier/api/subscriptions/addSubscription.ts:10]: Thank you, I also added it to deleteSubscription.ts 🙏

**Comment by CarinaWolli** [comment]
@zomars implemented your feedback, I would need your re-approval now 🙏

**Review by zomars — APPROVED** [review]

**Comment by zomars** [comment]
@shivamklr Can we test this PR on QA before merging? 🙏🏽 (or after we tag later today)

**Pull request merged**